Add CLI pages for hpopt and fingerprint

jonwzheng opened the pull request:
This PR adds documentation for the `hpopt` and `fingerprint` commands (addresses #815). 

I worked on hpopt while Joel worked on fingerprinting.

shihchengli commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,74 @@
+
+Note that the hyperparameter optimization script sees all the data given to it. The intended use is to run the hyperparameter optimization script on a dataset with the eventual test set held out. 
+
+.. If you need to optimize hyperparameters separately for several different cross validation splits, you should e.g. set up a bash script to run :code:`chemprop hpopt` separately on each split's training and validation data with test held out.
```
```suggestion
.. note::
    If you need to optimize hyperparameters separately for several different cross validation splits, you should e.g. set up a bash script to run :code:`chemprop hpopt` separately on each split's training and validation data with test held out.
```

jonwzheng replied:
I had left this commented out from v1 as I wasn't sure if this was still the recommended method, is this still correct?

shihchengli replied:
After reviewing the code, we use only the first split to perform HPO when `--num-folds` is greater than 1 (see [here](https://github.com/chemprop/chemprop/blob/main/chemprop/cli/hpopt.py#L485-L486)). Therefore, if one wants to optimize hyperparameters across different splits, they should provide their own data splits.
@hwpang we should add a sentence in the documentation and also a message in the CLI to let people know that only the first split is used.

jonwzheng replied:
okay, will add this as well

shihchengli commented on docs/source/tutorial/cli/fingerprint.rst:
```diff
@@ -0,0 +1,36 @@
+
+.. code-block::
+  
+    chemprop predict --test-path tests/data/smis.csv \
+        --model-path tests/data/example_model_v2_regression_mol.ckpt \
+        --output fps.csv
```
```suggestion
    chemprop fingerprint --test-path tests/data/smis.csv \
    --model-path tests/data/example_model_v2_regression_mol.ckpt \
    --output fps.csv
```

shihchengli replied:
I am trying to make this consistent with what we used in the training documentation. If you think it makes sense, please also change the one in the prediction documentation.

shihchengli commented on docs/source/tutorial/cli/fingerprint.rst:
```diff
@@ -0,0 +1,36 @@
+   
+   chemprop fingerprint --test-path <test_path> --model-path <model_path> 
+
+where :code:`<test_path>` is the path to the CSV file containing SMILES strings, and :code:`<model_path>` is the path to the trained model (.ckpt or .pt file) or a directory containing the trained model files. By default, predictions will be saved to the same directory as the test path. If desired, a different directory can be specified by using :code:`--output <path>`
```
```suggestion
where :code:`<test_path>` is the path to the CSV file containing SMILES strings, and :code:`<model_path>` is the location of checkpoint(s) or model file(s) to use for prediction. It can be a path to either a single pretrained model checkpoint (.ckpt) or single pretrained model file (.pt), a directory that contains these files, or a list of path(s) and directory(s). By default, predictions will be saved to the same directory as the test path. If desired, a different directory can be specified by using :code:`--output <path>`
```

hwpang replied:
We should also mention that when a directory is provided, it only includes the model file (.pt)

shihchengli replied:
@hwpang I agree and I just edited the suggested code.

shihchengli reviewed (commented):
Thanks for creating these documentations!

jonwzheng commented on docs/source/tutorial/cli/index.rst:
```diff
@@ -42,18 +44,19 @@ For more details on each command, see the corresponding section below:
+* :ref:`fingerprint`
 
-The following features are not yet implemented, but will soon be included in a future release:
+The following features are not yet implemented, but may be included in a future release:
```
okay! are we okay with using this language then? ("may" instead of "will soon")?

shihchengli replied:
My personal opinion is not to use "may." For something that hasn't been implemented, we can simply say "not implemented yet" or "will be implemented," since "very soon" does not provide much context about when it will become available (we know it will happen in version 2.1, which is scheduled for October). In this case, we should probably just keep it as "will be."

jonwzheng commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,74 @@
+   --task-type regression \
+   --config-path results/best_config.toml \
+
+Note that the hyperparameter optimization script sees all the data given to it. The intended use is to run the hyperparameter optimization script on a dataset with the eventual test set held out. 
```
Thanks for catching this; this is a remnant of the v1 documentation, so it seems the behavior is different from v1. I will change the documentation to reflect this.

hwpang commented on docs/source/tutorial/cli/fingerprint.rst:
```diff
@@ -0,0 +1,34 @@
+   
+   chemprop fingerprint --test-path <test_path> --model-path <model_path> 
+
+where :code:`<test_path>` is the path to the CSV file containing SMILES strings, and :code:`<model_path>` is the location of checkpoint(s) or model file(s) to use for prediction. It can be a path to either a single pretrained model checkpoint (.ckpt) or single pretrained model file (.pt), a directory that contains these files, or a list of path(s) and directory(s). If a directory, will recursivel ...
```
```suggestion
where :code:`<test_path>` is the path to the CSV file containing SMILES strings, and :code:`<model_path>` is the location of checkpoint(s) or model file(s) to use for prediction. It can be a path to either a single pretrained model checkpoint (.ckpt) or single pretrained model file (.pt), a directory that contains these files, or a list of path(s) and directory(s). If a directory, will recursively search and predict on all found (.pt) models. By default, predictions will be saved to the same directory as the test path. If desired, a different directory can be specified by using :code:`--output <path>`. The output <path> can end with either .csv or .npz, and the output will be saved to the corresponding file type.
```

jonwzheng replied:
thanks. Is this change true also for `predict` (`preds-path`)?

hwpang commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,80 @@
+
+ * :code:`depth` The number of message passing steps
+ * :code:`ffn_num_layers` The number of layers in the FFN model
+ * :code:`dropout` The probability (from 0 to 1) of dropout in the MPNN & FNN layers
```
Not sure if this is necessary

```suggestion
 * :code:`dropout` The probability (from 0.0 to 1.0) of dropout in the MPNN & FNN layers
```

hwpang commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,80 @@
+Another option is :code:`learning_rate` which includes:
+
+ * :code:`max_lr` The maximum learning rate
+ * :code:`init_lr_ratio` Ratio of the initial learning rate to the max learning rate
+ * :code:`final_lr_ratio` Ratio of the final learning rate to the max learning rate 
```
In the CLI, we only expect the user to input either `init_lr` or `final_lr`

```suggestion
 * :code:`init_lr` The initial learning rate. It is searched as a ratio relative to the max learning rate
 * :code:`final_lr` The initial learning rate. It is searched as a ratio relative to the max learning rate 
```

hwpang commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,80 @@
+
+The following other common keywords may be used:
+ 
+ * :code:`--raytune-num-samples <num_samples>` The number of conditions to sample 
```
```suggestion
 * :code:`--raytune-num-samples <num_samples>` The number of trials to perform
```

hwpang commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,80 @@
+
+Splitting
+----------
+By default, Chemprop will split the data into train / validation / test data splits. The splitting behavior can be modified using the same splitting arguments used in training, i.e. section :ref:`train-validation-test-splits`.
```
```suggestion
By default, Chemprop will split the data into train / validation / test data splits. The splitting behavior can be modified using the same splitting arguments used in training, i.e., section :ref:`train-validation-test-splits`.
```

jonwzheng replied:
Thanks, will address this in a separate commit I authored

jonwzheng commented:
@hwpang @shihchengli Thanks for the reviews! I've addressed your comments. You can preview the docs [here (hpopt)](https://chemprop--914.org.readthedocs.build/en/914/tutorial/cli/hpopt.html) and [here (fingerprint)](https://chemprop--914.org.readthedocs.build/en/914/tutorial/cli/hpopt.html).

hwpang commented on docs/source/tutorial/cli/hpopt.rst:
```diff
@@ -0,0 +1,80 @@
+Applying Optimal Hyperparameters
+---------------------------------
+
+Once hyperparameter optimization is complete, the optimal hyperparameters can be applied during training by specifying the config path. For example:
+
+.. code-block::
+
+   chemprop train --data-path tests/data/regression.csv \
+   --task-type regression \
+   --config-path results/best_config.toml \
```
I think we don't need to add the task type here

```suggestion
Once hyperparameter optimization is complete, the optimal hyperparameters can be applied during training by specifying the config path. If an argument is both provided via the command line and the config file, the command line takes precedence. For example:

.. code-block::

   chemprop train --data-path tests/data/regression.csv \
   --config-path results/best_config.toml \
```

hwpang reviewed (commented):
LGTM otherwise!
